### PR TITLE
[BEAM-7700] Java transform catalog

### PR DESCRIPTION
**Pull request by rosetn**
Java transform catalog

Java version of the Python transform catalog [BEAM-7464](https://github.com/apache/beam/pull/8534)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by rosetn** [comment]
R: @robinyqiu PTAL

STAGED:
http://apache-beam-website-pull-requests.storage.googleapis.com/9129/documentation/transforms/java/overview/index.html

File: website/src/documentation/transforms/java/other/view.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    </a>
+</table>
+<br>
+Operations for turning a collection into view that may be used as a side-input to a `ParDo`.
+
+## Examples
+See [BEAM-7704](https://issues.apache.org/jira/browse/BEAM-7704) for updates. 
+
+## Related transforms 
+* [GroupByKey]({{ site.baseurl }}/documentation/transforms/java/aggregation/groupbykey)
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/other/view.md]: Refer to `ParDo` and `CombineWithContext` since they are able to consume `PCollectionView`, instead of these 2 transforms.

File: website/src/documentation/transforms/java/other/window.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+</table>
+<br>
+Logically divides up or groups the elements of a collection into finite
+windows according to a function.
+
+## Examples
+See [BEAM-7704](https://issues.apache.org/jira/browse/BEAM-7704) for updates.
+
+## Related transforms 
+* [GroupByKey]({{ site.baseurl }}/documentation/transforms/java/aggregation/groupbykey)
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/other/window.md]: Change this with `Reify`?

File: website/src/documentation/transforms/java/aggregation/cogroupbykey.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    Iterable<Integer> allIntegers = result.getAll(t1);
+    // Retrieve the string associated with this key from pt2.
+    // Note: This will fail if multiple values had the same key in pt2.
+    String string = e.getOnly(t2);
+    ...
+}));
+```
+
+## Related transforms 
+* [CombineWithContext]({{ site.baseurl }}/documentation/transforms/java/aggregation/combinewithcontext) to combine elements.
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/cogroupbykey.md]: This doesn't seem related.

File: website/src/documentation/transforms/java/aggregation/combine.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+     Javadoc
+    </a>
+</table>
+<br>
+A user-defined `CombineFn` may be applied to combine all elements in a
+`PCollection` (global combine) or to combine all elements associated
+with each key. 
+
+While the result is similar to applying a `GroupByKey` followed by
+counting the number of values in each `Iterable`, there is an impact
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/combine.md]: I don't think it is only "counting the number of". "aggregating" sounds more correct to me. Same below.

File: website/src/documentation/transforms/java/aggregation/distinct.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+Does not preserve the order of the input collection.
+
+On some data sets, it might be more efficient to compute an approximate
+answer using `ApproximateUnique`, which also allows for determining distinct
+values for each key.
+
+## Examples
+See [BEAM-7703](https://issues.apache.org/jira/browse/BEAM-7703) for updates.
+
+## Related transforms
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/distinct.md]: Add reference to `ApproximateUnique`?

File: website/src/documentation/transforms/java/aggregation/distinct.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        href="https://beam.apache.org/releases/javadoc/current/index.html?org/apache/beam/sdk/transforms/Distinct.html">
+      <img src="https://beam.apache.org/images/logos/sdks/java.png" width="20px" height="20px"
+           alt="Javadoc" />
+     Javadoc
+    </a>
+</table>
+<br>
+Produces a collection containing distinct elements of the input collection.
+
+Does not preserve the order of the input collection.
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/distinct.md]: Why calling out this here? I think `PCollection` is unordered and most transforms do not have any guarantee on ordering.

File: website/src/documentation/transforms/java/aggregation/groupbykey.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      <img src="https://beam.apache.org/images/logos/sdks/java.png" width="20px" height="20px"
+           alt="Javadoc" />
+     Javadoc
+    </a>
+</table>
+<br>
+Takes a keyed collection of elements and produces a collection where
+each element consists of a key and all values associated with that key.
+
+Normally, the result of a `GroupByKey` contains one element for each key
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/groupbykey.md]: This sentence might be confusing. Even with windowing and triggering, each time the output is still a (k, Iterable) pair. I would just remove this sentence, and update the one above to "... consists of a key and a `Iterable` of all values associated with that key."

File: website/src/documentation/transforms/java/aggregation/latest.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+**Example**: compute the latest value for each session
+```java
+ PCollection input = ...;
+ PCollection sessioned = input
+    .apply(Window.into(Sessions.withGapDuration(Duration.standardMinutes(5)));
+ PCollection latestValues = sessioned.apply(Latest.globally());
+```
+
+## Related transforms 
+* [Sample]({{ site.baseurl }}/documentation/transforms/java/aggregation/sample)
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/latest.md]: Refer to `WithTimestamp` and `Reify` instead.

File: website/src/documentation/transforms/java/index.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  <tr><td><a href="{{ site.baseurl }}/documentation/transforms/java/aggregation/distinct">Distinct</a></td><td>Produces a collection containing distinct elements from the input collection.</td></tr>  
+  <tr><td><a href="{{ site.baseurl }}/documentation/transforms/java/aggregation/groupbykey">GroupByKey</a></td><td>Takes a keyed collection of elements and produces a collection where each element 
+  consists of a key and all values associated with that key.</td></tr>
+  <tr><td><a href="{{ site.baseurl }}/documentation/transforms/java/aggregation/groupintobatches">GroupIntoBatches</a></td><td>Batches values associated with keys into <code>Iterable</code> batches of some size. Each batch contains elements associated with a specific key.</td></tr>
+  <tr><td><a href="{{ site.baseurl }}/documentation/transforms/java/aggregation/latest">Latest</a></td><td>Selects the latest element within each aggregation according to the implicit timestamp.</td></tr>
+  <tr><td><a href="{{ site.baseurl }}/documentation/transforms/java/aggregation/mean">Max</a></td><td>Outputs the maximum element within each aggregation.</td></tr>
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/index.md]: Wrong link to mean.md

File: website/src/documentation/transforms/java/aggregation/max.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+```java
+ PCollection input = ...;
+ PCollection max = input.apply(Max.doublesGlobally());
+```
+
+**Example 2**: calculate the maximum of the `Integers` associated
+with each unique key (which is of type `String`).
+
+```java
+ PCollection> input = ...;
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/max.md]: Fix typo and add types here. Same for mean.md and sum.md

File: website/src/documentation/transforms/java/aggregation/max.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+</table>
+<br>
+Provides a variety of different transforms for computing the maximum
+values in a collection, either globally or for each key.
+
+## Examples
+**Example 1**: get the maximum of a `PCollection` of `Doubles`.
+
+```java
+ PCollection input = ...;
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/max.md]: `PCollection<Double>`

File: website/src/documentation/transforms/java/aggregation/mean.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+           alt="Javadoc" />
+     Javadoc
+    </a>
+</table>
+<br>
+Transforms for computing the arithmetic mean of the elements in a collection,
+or the mean of the values associated with each key in a collection of key-value pairs.
+
+* `Mean.globally()` returns a transform that then returns a collection whose contents is the mean of the input collection's elements. If there are no elements in the input collection, it returns 0.
+* `Mean.of()` is a `Combine.CombineFn` that computes the arithmetic mean of the elements in a collection.
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/mean.md]: Users don't have to use this `combineFn` directly. So I think this can be removed from here.

**Comment by aaltay** [comment]
Thank you for cc'ing me. No I have not seen the PR before, although I have seen the drafts for it. It looks great.

File: website/src/documentation/transforms/java/aggregation/combine.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    new Sum.SumDoubleFn()));
+// The combined value is of a different type than the original collection of values per key. PCollection has
+// keys of type String and values of type Integer, and the combined value is a Double.
+PCollection<KV<String, Integer>> playerAccuracy = ...;
+PCollection<KV<String, Double>> avgAccuracyPerPlayer =
+  playerAccuracy.apply(Combine.<String, Integer, Double>perKey(
+    new MeanInts())));
+```
+
+```java
```
[the remark this answers is not in the document]
Comment (rosetn) [reply website/src/documentation/transforms/java/aggregation/combine.md]: Did not catch that. Thanks!

File: website/src/documentation/transforms/java/aggregation/mean.md (hunk 12 above)
Comment (rosetn) [reply website/src/documentation/transforms/java/aggregation/mean.md]: Do you mean remove "is a `Combine.CombineFn` that" ?

**Comment by rosetn** [comment]
Thanks! Made the suggested edits; there's one open comment.

Tracking Python recommendations in [BEAM-7706](https://issues.apache.org/jira/browse/BEAM-7706)

I agree it is not as visible, but right now it is mean to be for "reference." Let's revisit this when we pull in Colab examples, so that it serves more as "learning."

We originally wanted to replace the Beam programming guide Section 4.2  with this, but some folks prefer the one-page reference style. However, the transform catalogs have pointers back to Section 4.2 when appropriate. 

Adding I/O transforms and possibly schema-aware ones sounds great. I can open a PR for that once and plan for later work if it's confirmed that the latter is stable enough to document.

File: website/src/documentation/transforms/java/aggregation/mean.md (hunk 12 above)
Comment (robinyqiu) [reply website/src/documentation/transforms/java/aggregation/mean.md]: No. I mean remove this sentence about `Mean.of()` entirely. Because it is not really a `PTransform`.

File: website/src/documentation/transforms/java/aggregation/max.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+```java
+PCollection<Double> input = ...;
+PCollection<Double> max = input.apply(Max.doublesGlobally());
+```
+
+**Example 2**: calculate the maximum of the `Integers` associated
+with each unique key (which is of type `String`).
+
+```java
+PCollection<Double> input = ...;
```
Comment (robinyqiu) [inline website/src/documentation/transforms/java/aggregation/max.md]: Typo here and for Mean and Sum: `Double` should be `KV<String, Integer>`, according to the description above.

Comment (rosetn) [reply website/src/documentation/transforms/java/aggregation/max.md]: Thanks!

**Comment by rosetn** [comment]
Thanks! Updated according to comments.

**Comment by robinyqiu** [comment]
Hi @rosetn, thanks for the changes. Have you committed all your changes? It seems there are some comments left unaddressed. You may need to click "load hidden comments" to see the full comment list. Thanks!

File: website/src/documentation/transforms/java/element-wise/partition.md [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+// You can extract each partition from the PCollectionList using the get method, as follows:
+PCollection<Student> fortiethPercentile = studentsByPercentile.get(4);
+```
+
+## Related transforms 
+* [Filter]({{ site.baseurl }}/documentation/transforms/java/elementwise/filter) is useful if the function is just 
+  deciding whether to output an element or not.
+* [ParDo]({{ site.baseurl }}/documentation/transforms/java/elementwise/pardo) is the most general element-wise mapping
+  operation, and includes other abilities such as multiple output collections and side-inputs. 
+* [CoGroupByKey]({{ site.baseurl }}/documentation/transforms/java/aggregation/cogroupbykey)
```
[the remark this answers is not in the document]
Comment (rosetn) [reply website/src/documentation/transforms/java/element-wise/partition.md]: From DPE @davidcavazos "It's kind of the opposite operation. Partition splits into multiple PCollections, while you can join them back together with CoGroupByKey, so they might be used complementary"

[the remark this answers is not in the document]
Comment (robinyqiu) [reply website/src/documentation/transforms/java/element-wise/partition.md]: I see. That's fair.

File: website/src/documentation/transforms/java/element-wise/filter.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+```java
+PCollection<Long> numbers = Create.of(1L, 2L, 3L, 4L, 5L);
+PCollection<Long> bigNumbers = numbers.apply(Filter.greaterThan(3));
+PCollection<Long> smallNumbers = numbers.apply(Filter.lessThanEq(3));
+```
+Other variants include `Filter.greaterThanEq`, `Filter.lessThan` and `Filter.equal`.
+
+## Related transforms 
+* [FlatMapElements]({{ site.baseurl }}/documentation/transforms/java/elementwise/flatmapelements) behaves the same as `Map`, but for
```
[the remark this answers is not in the document]
Comment (rosetn) [reply website/src/documentation/transforms/java/element-wise/filter.md]: @davidcavazos 's reasoning: "You can create a Filter by using FlatMapElements. Meaning that Filter is a subset of FlatMap, something similar as to why FlatMap and ParDo are related."

File: website/src/documentation/transforms/java/element-wise/filter.md [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+```java
+PCollection<String> allStrings = Create.of("Hello", "world", "hi");
+PCollection<String> longStrings = allStrings
+    .apply(Filter.by(new SerializableFunction<String, Boolean>() {
+      @Override
+      public Boolean apply(String input) {
+        return input.length() > 3;
+      }
+    }));
+```
```
[the remark this answers is not in the document]
Comment (rosetn) [reply website/src/documentation/transforms/java/element-wise/filter.md]: I agree. But @davidcavazos is replacing these samples with his colab in the new few months, so it might make more sense to invest time into that.

These snippets were written by @kennknowles and @bjchambers a while ago.

**Comment by rosetn** [comment]
I didn't see those other comments. Thanks, @robinyqiu! I made the updates.

**Comment by robinyqiu** [comment]
LGTM. Thanks for the changes!

**Comment by pabloem** [comment]
Thanks Rose! Thanks for reviewing Robin!
